packages/protocol/contracts/dapis/RrpDapiServer.sol

```diff
@@ -0,0 +1,178 @@
+        uint64 nextRequestIndex;
+        uint64 requestIndexResetCount;
+        address requestIndexResetter;
+        }
+    
+    struct DapiRequestIdentifiers {
+        bytes16 dapiId;
+        uint64 dapiRequestIndex;
+        uint64 dapiRequestIndexResetCountAtRequestTime;
+        }
+    
+    AirnodeRrp public airnodeRrp;
+    mapping(bytes16 => Dapi) private dapis;
+    uint256 public nextDapiIndex = 1;
+    mapping(bytes32 => DapiRequestIdentifiers) private airnodeRequestIdToDapiRequestIdentifiers;
+
+    constructor(address _airnodeRrp) {
+        airnodeRrp = AirnodeRrp(_airnodeRrp);
+    }
+
+    function registerDapi(
+        uint256 noResponsesToReduce,
+        int256 toleranceInPercentages,
+        uint256 requesterIndex,
+        bytes32[] calldata templateIds,
+        address[] calldata designatedWallets,
+        address reduceAddress,
+        bytes4 reduceFunctionId,
+        address requestIndexResetter
+        )
+        external
+        returns (bytes16 dapiId)
+    {
+      require(
+          templateIds.length == designatedWallets.length,
+          "Parameter lengths do not match"
+          );
+      require(
+          noResponsesToReduce <= templateIds.length && noResponsesToReduce != 0,
+          "Invalid number of responses to reduce"
+          );
+      dapiId = bytes16(keccak256(abi.encodePacked(
+          noResponsesToReduce,
+          toleranceInPercentages,
+          requesterIndex,
+          templateIds,
+          designatedWallets,
+          reduceAddress,
+          reduceFunctionId,
+          requestIndexResetter
+          )));
+      dapis[dapiId].noResponsesToReduce = noResponsesToReduce;
+      dapis[dapiId].toleranceInPercentages = toleranceInPercentages;
+      dapis[dapiId].requesterIndex = requesterIndex;
+      dapis[dapiId].templateIds = templateIds;
+      dapis[dapiId].designatedWallets = designatedWallets;
+      dapis[dapiId].reduceAddress = reduceAddress;
+      dapis[dapiId].reduceFunctionId = reduceFunctionId;
+      dapis[dapiId].requestIndexResetter = requestIndexResetter;
+      dapis[dapiId].nextRequestIndex = 1;
+    }
+
+    function resetDapiRequestIndex(bytes16 dapiId)
+        external
+    {
+        Dapi storage dapi = dapis[dapiId];
+        require(
+            msg.sender == dapi.requestIndexResetter,
+            "Caller not resetter"
+            );
+        dapi.nextRequestIndex = 1;
+        dapi.requestIndexResetCount++;
+    }
+
+    function makeDapiRequest(
+        bytes16 dapiId,
+        bytes calldata parameters
+        )
+        external
+        returns (uint64 currDapiRequestIndex)
+    {
+        Dapi storage dapi = dapis[dapiId];
+        currDapiRequestIndex = dapi.nextRequestIndex++;
+        dapi.requestIndexToResponsesLength[currDapiRequestIndex] = 0;
+        if (dapi.requestIndexToResponses[currDapiRequestIndex].length == 0) {
+            dapi.requestIndexToResponses[currDapiRequestIndex] = new int256[](dapi.noResponsesToReduce);
```

Comment:
Do you have to create a new array? Can't you just reinsert into the same array?

Reply:
Oh, wait, that is what's happening right? You allocate an array of length `dapi.noResponsesToReduce` and then just rewrite to it?

Reply:
Yep